net/xia/ppal_ether/main.c

```diff
@@ -271,6 +272,36 @@ static void detach_neigh_to_interface(struct fib_xid_ether_main *mether)
 	ether_interface_put(einterface);
 }
 
+/* Initialize the cached Ethernet header. */
+static void xia_ether_header_cache(struct fib_xid_ether_main *mfxid)
+{
+	struct ethhdr *eth;
+	const struct net_device *dev = mfxid->host_interface;
+
+	seqlock_init(&mfxid->cached_hdr.hh_lock);
+	eth = (struct ethhdr *)
+	    (((u8 *)mfxid->cached_hdr.hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+
+	eth->h_proto = htons(ETH_P_XIP);
+	memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
+	memcpy(eth->h_dest, mfxid->ha, ETH_ALEN);
+	mfxid->cached_hdr.hh_len = ETH_HLEN;
+}
+
+/* Update the cached header with new Ethernet address. */
+static void xia_ether_header_cache_update(struct fib_xid_ether_main *mfxid)
+{
+	struct ethhdr *eth;
+	const struct net_device *dev = mfxid->host_interface;
+
+	eth = (struct ethhdr *)
+	    (((u8 *)mfxid->cached_hdr.hh_data) + (HH_DATA_OFF(sizeof(*eth))));
+
+	write_seqlock_bh(&mfxid->cached_hdr.hh_lock);
+	memcpy(eth->h_source, dev->dev_addr, ETH_ALEN);
```

Comment:
Replace constant `ETH_ALEN` with `dev->addr_len`.

Reply:
Ignore all this comment. Since we are copying to `eth`, which has a fixed size, the current code is safer.